haystack_experimental/dataclasses/tool.py

```diff
@@ -85,17 +95,117 @@ def to_dict(self) -> Dict[str, Any]:
+            Deserialized Tool.
         """
         data["function"] = deserialize_callable(data["function"])
         return cls(**data)
 
+    @classmethod
+    def from_function(cls, function: Callable, use_docstring_as_tool_description: bool = True) -> "Tool":
+        """
+        Create a Tool instance from a function.
+
+        Usage example:
+        ```python
+        from typing import Annotated, Literal
+        from haystack_experimental.dataclasses import Tool
+
+        def get_weather(
+            city: Annotated[str, "the city for which to get the weather"] = "Munich",
+            unit: Annotated[Literal["Celsius", "Fahrenheit"], "the unit for the temperature"] = "Celsius"):
+            '''A simple function to get the current weather for a location.'''
+            return f"Weather report for {city}: 20 {unit}, sunny"
+
+        tool = Tool.from_function(get_weather)
+
+        print(tool)
+        >>> Tool(name='get_weather', description='A simple function to get the current weather for a location.',
+        >>> parameters={
+        >>> 'type': 'object',
+        >>> 'properties': {
+        >>>     'city': {'type': 'string', 'description': 'the city for which to get the weather', 'default': 'Munich'},
+        >>>     'unit': {
+        >>>         'type': 'string',
+        >>>         'enum': ['Celsius', 'Fahrenheit'],
+        >>>         'description': 'the unit for the temperature',
+        >>>         'default': 'Celsius',
+        >>>     },
+        >>>     }
+        >>> },
+        >>> function=<function get_weather at 0x7f7b3a8a9b80>)
+        ```
+
+        :param function:
+            The function to be converted into a Tool.
+            The function must include type hints for all parameters.
+            If a parameter is annotated using `typing.Annotated`, its metadata will be used as parameter description.
+        :param use_docstring_as_tool_description:
+            Whether to use the function's docstring as the tool description.
+
+        :returns:
+            The Tool created from the function.
+
+        :raises ValueError:
+            If any parameter of the function lacks a type hint.
+        :raises SchemaGenerationError:
+            If there is an error generating the JSON schema for the Tool.
+        """
+        tool_description = ""
+        if use_docstring_as_tool_description and function.__doc__:
+            tool_description = function.__doc__
+
+        signature = inspect.signature(function)
+
+        # collect fields (types and defaults) and descriptions from function parameters
+        fields: Dict[str, Any] = {}
+        descriptions = {}
+
+        for name, param in signature.parameters.items():
+            if param.annotation is param.empty:
+                raise ValueError(f"Function '{function.__name__}': parameter '{name}' does not have a type hint.")
+            default = param.default if param.default is not param.empty else ...
```

Comment:
Let's add a comment to explain why we need ellipses here.

Reply:
done